README.markdown
Node DDP Client
===============

A callback style [DDP](https://github.com/meteor/meteor/blob/devel/packages/livedata/DDP.md) ([Meteor](http://meteor.com/)'s Distributed Data Protocol) node client, originally based alansikora's [node-js_ddp-client](https://github.com/alansikora/node-js_ddp-client) and Meteor's python client. Uses a more callback style approach.

The client implements version 1 of DDP, as well as fallbacks to pre1 and pre2.

Installation
============

```
  npm install ddp
```

Authentication
==============
Built-in authentication support was removed in ddp 0.7.0 due to changes in Meteor version 0.8.2.

DDP Authentication is now implemented by [vsivsi/ddp-login](https://github.com/vsivsi/ddp-login).

A quick and dirty (but insecure) alternative is to use plain-text logins via a method call:

```js
// logging in with e-mail
ddpclient.call("login", [
  { user : { email : "user@domain.com" }, password : "password" }
], function (err, result) { ... });

// logging in with username
ddpclient.call("login", [
  { user : { username : "username" }, password : "password" }
], function (err, result) { ... });
```


Example
=======

Please see the example in `examples/example.js`. Or here for reference:

```js
var DDPClient = require("ddp");

var ddpclient = new DDPClient({
  // All properties optional, defaults shown
  host : "localhost",
  port : 3000,
  ssl  : false,
  autoReconnect : true,
  autoReconnectTimer : 500,
  maintainCollections : true,
  ddpVersion : '1',  // ['1', 'pre2', 'pre1'] available
  // uses the SockJs protocol to create the connection
  // this still uses websockets, but allows to get the benefits
  // from projects like meteorhacks:cluster
  // (for load balancing and service discovery)
  // do not use `path` option when you are using useSockJs
  useSockJs: true,
  // Use a full url instead of a set of `host`, `port` and `ssl`
  // do not set `useSockJs` option if `url` is used
  url: 'wss://example.com/websocket'
});

/*
 * Connect to the Meteor Server
 */
ddpclient.connect(function(error, wasReconnect) {
  // If autoReconnect is true, this callback will be invoked each time
  // a server connection is re-established
  if (error) {
    console.log('DDP connection error!');
    return;
  }

  if (wasReconnect) {
    console.log('Reestablishment of a connection.');
  }

  console.log('connected!');

  setTimeout(function () {
    /*
     * Call a Meteor Method
     */
    ddpclient.call(
      'deletePosts',             // name of Meteor Method being called
      ['foo', 'bar'],            // parameters to send to Meteor Method
      function (err, result) {   // callback which returns the method call results
        console.log('called function, result: ' + result);
      },
      function () {              // callback which fires when server has finished
        console.log('updated');  // sending any updated documents as a result of
        console.log(ddpclient.collections.posts);  // calling this method
      }
    );
  }, 3000);

  /*
   * Call a Meteor Method while passing in a random seed.
   * Added in DDP pre2, the random seed will be used on the server to generate
   * repeatable IDs. This allows the same id to be generated on the client and server
   */
  var Random = require("ddp-random"),
      random = Random.createWithSeeds("randomSeed");  // seed an id generator

  ddpclient.callWithRandomSeed(
    'createPost',              // name of Meteor Method being called
    [{ _id : random.id(),      // generate the id on the client
      body : "asdf" }],
    "randomSeed",              // pass the same seed to the server
    function (err, result) {   // callback which returns the method call results
      console.log('called function, result: ' + result);
    },
    function () {              // callback which fires when server has finished
      console.log('updated');  // sending any updated documents as a result of
      console.log(ddpclient.collections.posts);  // calling this method
    }
  );

  /*
   * Subscribe to a Meteor Collection
   */
  ddpclient.subscribe(
    'posts',                  // name of Meteor Publish function to subscribe to
    [],                       // any parameters used by the Publish function
    function () {             // callback when the subscription is complete
      console.log('posts complete:');
      console.log(ddpclient.collections.posts);
    }
  );

  /*
   * Observe a collection.
   */
  var observer = ddpclient.observe("posts");
  observer.added = function(id) {
    console.log("[ADDED] to " + observer.name + ":  " + id);
  };
  observer.changed = function(id, oldFields, clearedFields, newFields) {
    console.log("[CHANGED] in " + observer.name + ":  " + id);
    console.log("[CHANGED] old field values: ", oldFields);
    console.log("[CHANGED] cleared fields: ", clearedFields);
    console.log("[CHANGED] new fields: ", newFields);
  };
  observer.removed = function(id, oldValue) {
    console.log("[REMOVED] in " + observer.name + ":  " + id);
    console.log("[REMOVED] previous value: ", oldValue);
  };
  setTimeout(function() { observer.stop() }, 6000);
});

/*
 * Useful for debugging and learning the ddp protocol
 */
ddpclient.on('message', function (msg) {
  console.log("ddp message: " + msg);
});

/*
 * Close the ddp connection. This will close the socket, removing it
 * from the event-loop, allowing your application to terminate gracefully
 */
ddpclient.close();

/*
 * If you need to do something specific on close or errors.
 * You can also disable autoReconnect and
 * call ddpclient.connect() when you are ready to re-connect.
*/
ddpclient.on('socket-close', function(code, message) {
  console.log("Close: %s %s", code, message);
});

ddpclient.on('socket-error', function(error) {
  console.log("Error: %j", error);
});

/*
 * You can access the EJSON object used by ddp.
 */
var oid = new ddpclient.EJSON.ObjectID();
```

SockJS Mode
===============

By using the `useSockJs` option like below, DDP connection will use [SockJs](https://github.com/sockjs) protocol to establish the WebSocket connection.

```js
var ddpClient = new DDPClient({ useSockJs: true });
```

Meteor server uses SockJs to implement it's DDP server. With this mode, we can get the benefits provided by [meteorhacks:cluster](https://github.com/meteorhacks/cluster). Some of those are load balancing and service discovery.

* For load balancing you don't need to anything.
* For service discovery, just use the `path` option to identify the service you are referring to.

> With this mode, `path` option has a special meaning. So, thing twice before using `path` option when you are using this option.

Unimplemented Features
====
The node DDP client does not implement ordered collections, something that while in the DDP spec has not been implemented in Meteor yet.

Thanks
======

Many thanks to Alan Sikora for the ddp-client which formed the inspiration for this code.

Contributions:
 * Tom Coleman (@tmeasday)
 * Thomas Sarlandie (@sarfata)
 * Mason Gravitt (@emgee3)
 * Mike Bannister (@possiblities)
 * Chris Mather (@eventedmind)
 * James Gill (@jagill)
 * Vaughn Iverson (@vsivsi)


CHANGELOG.md
0.12.2 — 2016-12-02

  - Fix ignored sends when server is disconnected (#90)
  - Pass data.fields as parameter to observer.added (#82)
  - Improve autoreconnect behavior (#63, see discussion)
  - Ensure connection status properly updated when optional connect function is omitted (#86)
  - Update npm dependencies

0.12.1 — 2016-09-15

  - Update npm dependencies, including request, to fix vulnerability (#89)

0.12.0 — 2016-03-31

  - Update npm dependencies, including faye-websocket (#78)
  - Catch JSON parsing exceptions (#77)
  - Fix adding changed handler on observer creator (#72)

0.11.0 — 2015-03-23

  - Allow passing url to websocket connection string (#52)
  - Allow passing TLS options to websocket connection (#53)
  - Track method calls so it incomplete calls can be handled (#54)

0.10.0 - 2015-02-03

  - Add optional SockJS support

0.9.4 - 2014-12-22

  - Handle socket errors occurring before connection is established

0.9.3 - 2014-12-20

  - Remove event listeners on close

0.9.2 - 2014-11-17

  - Observer callback emits changed fields.

0.9.1 - 2014-11-11

  - Collection elements now have _id field.

0.9.0 - 2014-11-05

  - Added collection observation

0.8.2 - 2014-10-11

  - Fixed variable name typo

0.8.1 - 2014-09-16

  - Remove debug console.logs

0.8.0 - 2014-09-16

  - Bump DDP version to 1
  - Fix connect callback handler (#41)
  - Change underscored_variable_names to camelCase

0.7.0 - 2014-06-23

  - Built-in support for authenticating to Meteor's Accounts system in has
    been removed, due to changes in Meteor's Accounts system in 0.8.2
    (https://github.com/meteor/meteor/blob/devel/History.md#meteor-accounts).
    If you need login support, try https://github.com/vsivsi/ddp-login
  - EJSON support now mandatory

0.6.0 - 2014-06-08

  - Update collection before emitting `message`.

0.5.2 - 2014-06-01

  - Added MIT License

0.5.1 - 2014-05-23

  - Switch to MDG-patched version of underscore.

0.5.0 - 2014-05-14

  - Use ddp-ejson instead of meteor-ejson. ddp-ejson is a repackage of
    Meteor's latest EJSON package
  - Use ddp-srp insead of node-srp. ddp-srp is a repackage of Meteor's
    latest SRP package
  - Added second callback to ddpclient.call, executed when the DDP
    `updated` message is received
  - Allow automatic EJSON serialization/deserialization of ObjectIDs
  - Expose EJSON package to allow for addition of custom EJSON types
  - added DDP pre2 support
    - DDP version negotiation
    - DDP heartbeat support (reply only)
    - `ddpclient.callWithRandomSeed` supports client-generated `_id`s

0.4.6 - 2014-04-28

  - Return id used when calling subscribe method

0.4.5 - 2014-04-24

  - Fix login with password method to return login token

0.4.4 - 2014-02-09

  - Fix a bug where if the server responded to an error on the first
    step of SRP authentication it was not handled correctly (i.e when
    the user is not found)

0.4.3 - 2013-12-19

  - Fix bug with socket reconnects tailspinning into an infinite loop
    (#30 by @jagill)
  - Fix bug when use_ejson was not always set properly by default.
    (#29 by @jagill)

0.4.2 - 2013-12-14

  - Use EJSON by default (#28)

0.4.1 - 2013-12-07

  - Ability to switch off collections monitoring

0.4.0 - 2013-12-07

  - Switched to faye-websockets (#26 by @jagill)

0.3.6 - 2013-11-07

  - fixed bug with default params when ignoring root certs (in case the
    machine doesn't have the cert)
  - Added DDP login with SRP authentication

0.3.5 - 2013-11-05

 - Added non strict SSL option in case of missing root certificates

0.3.4 - 2013-08-28

 - added EJSON support (default is off) with a couple tests

0.3.3 - 2013-05-29

 - fixed bug where an exception could be thrown when sending a message on
   a socket that is not opened anymore (issue #18)
 - added some tests (work in progress)

0.3.2 - 2013-04-08

  - fixed bug where client would reconnect when closing (@tarangp)

0.3.1 - 2013-04-06

  - added a failed message to the connect callback if version negotiation
    fails.

0.3.0 - 2013-03-18

  - moved over to DDP-pre1


examples/example-meteor-server/ddp-test.js
Posts = new Meteor.Collection('posts');

if (Meteor.isClient) {
  Template.posts.posts = function () {
    return Posts.find();
  };
}

if (Meteor.isServer) {
  Meteor.startup(function () {
    var postCount = Posts.find().count();
    if ( postCount < 10) {
      for ( ; postCount < 10;  postCount++) {
        Posts.insert({
          body : Random.secret()
        });
      }
    }
  });
}

Meteor.methods({
  deletePosts : function () {
    var cursor = Posts.find({}, { limit : 5 });
    cursor.forEach(function (post) {
      Posts.remove(post._id);
    });
  }
});


examples/example.js
"use strict";

var DDPClient = require("../lib/ddp-client");

var ddpclient = new DDPClient({
  // All properties optional, defaults shown
  host : "localhost",
  port : 3000,
  ssl  : false,
  autoReconnect : true,
  autoReconnectTimer : 500,
  maintainCollections : true,
  ddpVersion : "1",  // ["1", "pre2", "pre1"] available,
  // uses the sockJs protocol to create the connection
  // this still uses websockets, but allows to get the benefits
  // from projects like meteorhacks:cluster
  // (load balancing and service discovery)
  // do not use `path` option when you are using useSockJs
  useSockJs: true,
  // Use a full url instead of a set of `host`, `port` and `ssl`
  // do not set `useSockJs` option if `url` is used
  url: 'wss://example.com/websocket'
});

/*
 * Connect to the Meteor Server
 */
ddpclient.connect(function(error, wasReconnect) {
  // If autoReconnect is true, this callback will be invoked each time
  // a server connection is re-established
  if (error) {
    console.log("DDP connection error!");
    return;
  }

  if (wasReconnect) {
    console.log("Reestablishment of a connection.");
  }

  console.log("connected!");

  setTimeout(function () {
    /*
     * Call a Meteor Method
     */
    ddpclient.call(
      "deletePosts",             // name of Meteor Method being called
      ["foo", "bar"],            // parameters to send to Meteor Method
      function (err, result) {   // callback which returns the method call results
        console.log("called function, result: " + result);
      },
      function () {              // callback which fires when server has finished
        console.log("updated");  // sending any updated documents as a result of
        console.log(ddpclient.collections.posts);  // calling this method
      }
    );
  }, 3000);

  /*
   * Call a Meteor Method while passing in a random seed.
   * Added in DDP pre2, the random seed will be used on the server to generate
   * repeatable IDs. This allows the same id to be generated on the client and server
   */
  var Random = require("ddp-random"),
      random = Random.createWithSeeds("randomSeed");  // seed an id generator

  ddpclient.callWithRandomSeed(
    "createPost",              // name of Meteor Method being called
    [{ _id : random.id(),      // generate the id on the client
      body : "asdf" }],
    "randomSeed",              // pass the same seed to the server
    function (err, result) {   // callback which returns the method call results
      console.log("called function, result: " + result);
    },
    function () {              // callback which fires when server has finished
      console.log("updated");  // sending any updated documents as a result of
      console.log(ddpclient.collections.posts);  // calling this method
    }
  );

  /*
   * Subscribe to a Meteor Collection
   */
  ddpclient.subscribe(
    "posts",                  // name of Meteor Publish function to subscribe to
    [],                       // any parameters used by the Publish function
    function () {             // callback when the subscription is complete
      console.log("posts complete:");
      console.log(ddpclient.collections.posts);
    }
  );

  /*
   * Observe a collection.
   */
  var observer = ddpclient.observe("posts");
  observer.added = function(id) {
    console.log("[ADDED] to " + observer.name + ":  " + id);
  };
  observer.changed = function(id, oldFields, clearedFields) {
    console.log("[CHANGED] in " + observer.name + ":  " + id);
    console.log("[CHANGED] old field values: ", oldFields);
    console.log("[CHANGED] cleared fields: ", clearedFields);
  };
  observer.removed = function(id, oldValue) {
    console.log("[REMOVED] in " + observer.name + ":  " + id);
    console.log("[REMOVED] previous value: ", oldValue);
  };

  setTimeout(function() { observer.stop(); }, 6000);
});


lib/ddp-client.js
"use strict";

var util = require("util");
var events = require("events");
var WebSocket = require("faye-websocket");
var EJSON = require("ddp-ejson");
var request = require('request');
var pathJoin = require('path').join;
var _ = require("ddp-underscore-patched");

var DDPClient = function(opts) {
  var self = this;

  opts = opts || {};

  // backwards compatibility
  if ("use_ssl" in opts)              { opts.ssl = opts.use_ssl; }
  if ("auto_reconnect" in opts)       { opts.autoReconnect = opts.auto_reconnect; }
  if ("auto_reconnect_timer" in opts) { opts.autoReconnectTimer = opts.auto_reconnect_timer; }
  if ("maintain_collections" in opts) { opts.maintainCollections = opts.maintain_collections; }
  if ("ddp_version" in opts)          { opts.ddpVersion = opts.ddp_version; }

  // default arguments
  self.host = opts.host || "localhost";
  self.port = opts.port || 3000;
  self.path = opts.path;
  self.ssl  = opts.ssl  || self.port === 443;
  self.tlsOpts = opts.tlsOpts || {};
  self.useSockJs = opts.useSockJs || false;
  self.autoReconnect = ("autoReconnect" in opts) ? opts.autoReconnect : true;
  self.autoReconnectTimer = ("autoReconnectTimer" in opts) ? opts.autoReconnectTimer : 500;
  self.maintainCollections = ("maintainCollections" in opts) ? opts.maintainCollections : true;
  self.url  = opts.url;
  // support multiple ddp versions
  self.ddpVersion = ("ddpVersion" in opts) ? opts.ddpVersion : "1";
  self.supportedDdpVersions = ["1", "pre2", "pre1"];

  // Expose EJSON object, so client can use EJSON.addType(...)
  self.EJSON = EJSON;

  // very very simple collections (name -> [{id -> document}])
  if (self.maintainCollections) {
    self.collections = {};
  }

  // internal stuff to track callbacks
  self._isConnecting = false;
  self._isReconnecting = false;
  self._nextId = 0;
  self._callbacks = {};
  self._updatedCallbacks = {};
  self._pendingMethods = {};
  self._observers = {};
};


DDPClient.ERRORS = {
  DISCONNECTED: new Error("DDPClient: Disconnected from DDP server")
};


/**
 * Inherits from EventEmitter
 */
util.inherits(DDPClient, events.EventEmitter);


DDPClient.prototype._prepareHandlers = function() {
  var self = this;

  self.socket.on("open", function() {
    // just go ahead and open the connection on connect
    self._send({
      msg : "connect",
      version : self.ddpVersion,
      support : self.supportedDdpVersions
    });
  });

  self.socket.on("error", function(error) {
    // error received before connection was established
    if (self._isConnecting) {
      self.emit("failed", error.message);
    }

    self.emit("socket-error", error);
  });

  self.socket.on("close", function(event) {
    self.emit("socket-close", event.code, event.reason);
    self._endPendingMethodCalls();
    self._recoverNetworkError();
  });

  self.socket.on("message", function(event) {
    self._message(event.data);
    self.emit("message", event.data);
  });
};

DDPClient.prototype._clearReconnectTimeout = function() {
  var self = this;
  if (self.reconnectTimeout) {
    clearTimeout(self.reconnectTimeout);
    self.reconnectTimeout = null;
  }
};

DDPClient.prototype._recoverNetworkError = function() {
  var self = this;
  if (self.autoReconnect && ! self._isClosing) {
    self._clearReconnectTimeout();
    self.reconnectTimeout = setTimeout(function() { self.connect(); }, self.autoReconnectTimer);
    self._isReconnecting = true;
  }
};

///////////////////////////////////////////////////////////////////////////
// RAW, low level functions
DDPClient.prototype._send = function(data) {
  if (data.msg !== 'connect' && this._isConnecting) {
    this._endPendingMethodCalls()
  } else {
    this.socket.send(
      EJSON.stringify(data)
    );
  }
};

// handle a message from the server
DDPClient.prototype._message = function(data) {
  var self = this;

  data = EJSON.parse(data);

  // TODO: 'addedBefore' -- not yet implemented in Meteor
  // TODO: 'movedBefore' -- not yet implemented in Meteor

  if (!data.msg) {
    return;

  } else if (data.msg === "failed") {
    if (self.supportedDdpVersions.indexOf(data.version) !== -1) {
      this.ddpVersion = data.version;
      self.connect();
    } else {
      self.autoReconnect = false;
      self.emit("failed", "Cannot negotiate DDP version");
    }

  } else if (data.msg === "connected") {
    self.session = data.session;
    self.emit("connected");

  // method result
  } else if (data.msg === "result") {
    var cb = self._callbacks[data.id];

    if (cb) {
      cb(data.error, data.result);
      delete self._callbacks[data.id];
    }

  // method updated
  } else if (data.msg === "updated") {

    _.each(data.methods, function (method) {
      var cb = self._updatedCallbacks[method];
      if (cb) {
        cb();
        delete self._updatedCallbacks[method];
      }
    });

  // missing subscription
  } else if (data.msg === "nosub") {
    var cb = self._callbacks[data.id];

    if (cb) {
      cb(data.error);
      delete self._callbacks[data.id];
    }

  // add document to collection
  } else if (data.msg === "added") {
    if (self.maintainCollections && data.collection) {
      var name = data.collection, id = data.id;

      if (! self.collections[name])     { self.collections[name] = {}; }
      if (! self.collections[name][id]) { self.collections[name][id] = {}; }

      self.collections[name][id]._id = id;

      if (data.fields) {
        _.each(data.fields, function(value, key) {
          self.collections[name][id][key] = value;
        });
      }

      if (self._observers[name]) {
        _.each(self._observers[name], function(observer) {
          observer.added(id, data.fields);
        });
      }
    }

  // remove document from collection
  } else if (data.msg === "removed") {
    if (self.maintainCollections && data.collection) {
      var name = data.collection, id = data.id;

      if (! self.collections[name][id]) {
        return;
      }

      var oldValue = self.collections[name][id];

      delete self.collections[name][id];

      if (self._observers[name]) {
        _.each(self._observers[name], function(observer) {
          observer.removed(id, oldValue);
        });
      }
    }

  // change document in collection
  } else if (data.msg === "changed") {
    if (self.maintainCollections && data.collection) {
      var name = data.collection, id = data.id;

      if (! self.collections[name])     { return; }
      if (! self.collections[name][id]) { return; }

      var oldFields     = {},
          clearedFields = data.cleared || [],
          newFields = {};

      if (data.fields) {
        _.each(data.fields, function(value, key) {
            oldFields[key] = self.collections[name][id][key];
            newFields[key] = value;
            self.collections[name][id][key] = value;
        });
      }

      if (data.cleared) {
        _.each(data.cleared, function(value) {
            delete self.collections[name][id][value];
        });
      }

      if (self._observers[name]) {
        _.each(self._observers[name], function(observer) {
          observer.changed(id, oldFields, clearedFields, newFields);
        });
      }
    }

  // subscriptions ready
  } else if (data.msg === "ready") {
    _.each(data.subs, function(id) {
      var cb = self._callbacks[id];
      if (cb) {
        cb();
        delete self._callbacks[id];
      }
    });

  // minimal heartbeat response for ddp pre2
  } else if (data.msg === "ping") {
    self._send(
      _.has(data, "id") ? { msg : "pong", id : data.id } : { msg : "pong" }
    );
  }
};


DDPClient.prototype._getNextId = function() {
  return (this._nextId += 1).toString();
};


DDPClient.prototype._addObserver = function(observer) {
  if (! this._observers[observer.name]) {
    this._observers[observer.name] = {};
  }
  this._observers[observer.name][observer._id] = observer;
};


DDPClient.prototype._removeObserver = function(observer) {
  if (! this._observers[observer.name]) { return; }

  delete this._observers[observer.name][observer._id];
};

//////////////////////////////////////////////////////////////////////////
// USER functions -- use these to control the client

/* open the connection to the server
 *
 *  connected(): Called when the 'connected' message is received
 *               If autoReconnect is true (default), the callback will be
 *               called each time the connection is opened.
 */
DDPClient.prototype.connect = function(connected) {
  var self = this;
  self._isConnecting = true;
  self._connectionFailed = false;
  self._isClosing = false;

  self.addListener("connected", function() {
    self._clearReconnectTimeout();
    if (connected) {
      connected(undefined, self._isReconnecting);
    }
    self._isConnecting = false;
    self._isReconnecting = false;
  });
  self.addListener("failed", function(error) {
    self._isConnecting = false;
    self._connectionFailed = true;
    if (connected) {
      connected(error, self._isReconnecting);
    }
  });

  if (self.useSockJs) {
    self._makeSockJSConnection();
  } else {
    var url = self._buildWsUrl();
    self._makeWebSocketConnection(url);
  }
}

DDPClient.prototype._endPendingMethodCalls = function() {
  var self = this;
  var ids = _.keys(self._pendingMethods);
  self._pendingMethods = {};

  ids.forEach(function (id) {
    if (self._callbacks[id]) {
      self._callbacks[id](DDPClient.ERRORS.DISCONNECTED);
      delete self._callbacks[id];
    }

    if (self._updatedCallbacks[id]) {
      self._updatedCallbacks[id]();
      delete self._updatedCallbacks[id];
    }
  });
};

DDPClient.prototype._makeSockJSConnection = function() {
  var self = this;

  // do the info hit
  var protocol = self.ssl ? "https://" : "http://";
  var randomValue = "" + Math.ceil(Math.random() * 9999999);
  var path = pathJoin("/", self.path || "", "sockjs/info");
  var url = protocol + self.host + ":" + self.port + path;

  var requestOpts = { 'url': url, 'agentOptions': self.tlsOpts };

  request.get(requestOpts, function(err, res, body) {
    if (err) {
      self._recoverNetworkError();
    } else if (body) {
      var info;
      try {
        info = JSON.parse(body);
      } catch (e) {
        console.error(e);
      }
      if (!info || !info.base_url) {
        // no base_url, then use pure WS handling
        var url = self._buildWsUrl();
        self._makeWebSocketConnection(url);
      } else if (info.base_url.indexOf("http") === 0) {
        // base url for a different host
        var url = info.base_url + "/websocket";
        url = url.replace(/^http/, "ws");
        self._makeWebSocketConnection(url);
      } else {
        // base url for the same host
        var path = info.base_url + "/websocket";
        var url = self._buildWsUrl(path);
        self._makeWebSocketConnection(url);
      }
    } else {
      // no body. weird. use pure WS handling
      var url = self._buildWsUrl();
      self._makeWebSocketConnection(url);
    }
  });
};

DDPClient.prototype._buildWsUrl = function(path) {
  var self = this;
  var url;
  path = path || self.path || "websocket";
  var protocol = self.ssl ? "wss://" : "ws://";
  if (self.url && !self.useSockJs) {
    url = self.url;
  } else {
    url = protocol + self.host + ":" + self.port;
    url += (path.indexOf("/") === 0)? path : "/" + path;
  }
  return url;
};

DDPClient.prototype._makeWebSocketConnection = function(url) {
  var self = this;
  self.socket = new WebSocket.Client(url, null, self.tlsOpts);
  self._prepareHandlers();
};

DDPClient.prototype.close = function() {
  var self = this;
  self._isClosing = true;
  self.socket.close();
  self.removeAllListeners("connected");
  self.removeAllListeners("failed");
};


// call a method on the server,
//
// callback = function(err, result)
DDPClient.prototype.call = function(name, params, callback, updatedCallback) {
  var self = this;
  var id = self._getNextId();

  self._callbacks[id] = function () {
    delete self._pendingMethods[id];

    if (callback) {
      callback.apply(this, arguments);
    }
  };

  self._updatedCallbacks[id] = function () {
    delete self._pendingMethods[id];

    if (updatedCallback) {
      updatedCallback.apply(this, arguments);
    }
  };

  self._pendingMethods[id] = true;

  self._send({
    msg    : "method",
    id     : id,
    method : name,
    params : params
  });
};


DDPClient.prototype.callWithRandomSeed = function(name, params, randomSeed, callback, updatedCallback) {
  var self = this;
  var id = self._getNextId();

  if (callback) {
    self._callbacks[id] = callback;
  }

  if (updatedCallback) {
    self._updatedCallbacks[id] = updatedCallback;
  }

  self._send({
    msg        : "method",
    id         : id,
    method     : name,
    randomSeed : randomSeed,
    params     : params
  });
};

// open a subscription on the server, callback should handle on ready and nosub
DDPClient.prototype.subscribe = function(name, params, callback) {
  var self = this;
  var id = self._getNextId();

  if (callback) {
    self._callbacks[id] = callback;
  }

  self._send({
    msg    : "sub",
    id     : id,
    name   : name,
    params : params
  });

  return id;
};

DDPClient.prototype.unsubscribe = function(id) {
  var self = this;

  self._send({
    msg : "unsub",
    id  : id
  });
};

/**
 * Adds an observer to a collection and returns the observer.
 * Observation can be stopped by calling the stop() method on the observer.
 * Functions for added, changed and removed can be added to the observer
 * afterward.
 */
DDPClient.prototype.observe = function(name, added, changed, removed) {
  var self = this;
  var observer = {};
  var id = self._getNextId();

  // name, _id are immutable
  Object.defineProperty(observer, "name", {
    get: function() { return name; },
    enumerable: true
  });

  Object.defineProperty(observer, "_id", { get: function() { return id; }});

  observer.added   = added   || function(){};
  observer.changed = changed || function(){};
  observer.removed = removed || function(){};

  observer.stop = function() {
    self._removeObserver(observer);
  };

  self._addObserver(observer);

  return observer;
};

module.exports = DDPClient;
